test/e2e/framework/kcp.go

```diff
@@ -0,0 +1,223 @@
+package framework
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/spf13/pflag"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"github.com/kcp-dev/kcp/pkg/server"
+)
+
+// kcpServer exposes a kcp invocation to a test and
+// ensures the following semantics:
+//  - the command will get SIGINT 1 minutes before the test deadline
+//  - the command will get SIGKILL 10 seconds before the test deadline
+//  - unique etcd data dirs and ports for concurrent runs
+type kcpServer struct {
+	args        []string
+	ctx         context.Context
+	artifactDir string
+
+	lock *sync.Mutex
+	cfg  *rest.Config
+
+	t TestingTInterface
+}
+
+func (c *kcpServer) ArtifactDir() string {
+	return c.artifactDir
+}
+
+func newKcpServer(t *T, args ...string) (*kcpServer, error) {
+	t.Helper()
+	ctx := context.Background()
+	if deadline, ok := t.Deadline(); ok {
+		c, cancel := context.WithDeadline(ctx, deadline.Add(-10*time.Second))
+		ctx = c
+		t.Cleanup(cancel) // this does not really matter but govet is upset
+	}
+	kcpListenPort, err := GetFreePort(t)
+	if err != nil {
+		return nil, err
+	}
+	etcdClientPort, err := GetFreePort(t)
+	if err != nil {
+		return nil, err
+	}
+	etcdPeerPort, err := GetFreePort(t)
+	if err != nil {
+		return nil, err
+	}
+	artifactDir, err := ArtifactDir(t, kcpListenPort)
+	if err != nil {
+		return nil, err
+	}
+	return &kcpServer{
+		args: append([]string{
+			"--root_directory=" + artifactDir,
+			"--listen=:" + kcpListenPort,
+			"--etcd_client_port=" + etcdClientPort,
+			"--etcd_peer_port=" + etcdPeerPort,
+			"--kubeconfig_path=admin.kubeconfig"},
+			args...),
+		artifactDir: artifactDir,
+		ctx:         ctx,
+		t:           t,
+		lock:        &sync.Mutex{},
+	}, nil
+}
+
+// Run runs the kcp server while the parent context is active. This call is not blocking,
+// callers should ensure that the server is Ready() before using it.
+func (c *kcpServer) Run(parentCtx context.Context) error {
```

Comment:
:+1: good framework.

Btw, I notice that your out-of-process design has a nice side effect: it is hard to work around the flag interface. We are a small group now, but in upstream you have to carefully watch and review every PR against such a package because people like to add their own special tweak to the test server.